tc/tc_util.c

```diff
@@ -205,13 +207,18 @@ int parse_percent_rate(char *rate, const char *str, char *dev)
 	if (read_prop(dev, "speed", &max_rate_bits))
 		return -1;
 
-	ret = sscanf(str, "%lf%%", &perc);
-	if (ret != 1) {
-		fprintf(stderr, "Specified rate value could not be read or is malformed\n");
-		return -1;
-	}
+	ret = sscanf(str, "%m[0-9.%]", &char_perc);
+	if (ret != 1)
+		goto malf;
+
+	perc = strtod(char_perc, &p);
+	if (*p != '%' || *(p++) != '\0')
+		goto malf;
+
+	free(char_perc);
 
-	if (perc > 100.0 || perc < 0.0) {
+
+	if (perc > 100.0 || perc < 0.0 || errno == ERANGE) {
```

Comment:
This check for `errno == ERANGE` is for `strtod()`, but we call `free()` between the call to `strtod()` and here. It's possible that in some implementations `free()` could set `errno`.

So let's do the error checking for `strtod()` in this way:

    int saved_errno;

    ...

    perc = strtod(char_perc, &p);
    saved_errno = errno;
    free(str_perc);

    /* Make sure there's only one percent sign and it's at the end. */
    if (*p != '%' || *(p++) != '\0')
    	goto malf;
    if (perc > 100.0 || perc < 0.0 || saved_errno == ERANGE) {
    	fprintf(stderr, "Invalid rate specified; should be between [0,100]%% but is %s\n", str);
    	return -1;
    }

Reply:
This should catch any cases where there are multiple percentage signs, or a percentage sign that's not at the end. Please test these cases.

Reply:
Will do.

Reply:
Actually, the current code catches double percentage signs as I'm testing it right now, I guess `strtod()` set `*p` to where it stopped converting the string to double, and so, the other `(*p++) != '\0'` condition catches double percentage signs

Reply:
Right, sorry, I should have been clearer -- the code you wrote before already did that, I just added the comment to make it clear what it was doing.

But we definitely need to save `errno` in a variable so that we can use it after. Doing so also simplifies how we can can call `free(str_perc)`. So please add those parts in as well, following the suggested code that I wrote above.

Reply:
Right I see, thanks, will make the changes now